airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/UpdateConnectorConfigHelper.java

```diff
@@ -0,0 +1,115 @@
+import io.airbyte.api.client.model.generated.SourceRead;
+import io.airbyte.api.client.model.generated.SourceUpdate;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.protocol.models.Config;
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helper class for workers to persist updates to Source/Destination configs emitted from
+ * AirbyteControlMessages.
+ *
+ * This is in order to support connectors updating configs when running commands, which is specially
+ * useful for migrating configuration to a new version or for enabling connectors that require
+ * single-use or short-lived OAuth tokens.
+ */
+public class UpdateConnectorConfigHelper {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpdateConnectorConfigHelper.class);
+
+  private final AirbyteApiClient apiClient;
+
+  public UpdateConnectorConfigHelper(final AirbyteApiClient apiClient) {
+    this.apiClient = apiClient;
+  }
+
+  private UUID getConnectionIdFromJobId(final Long jobId) {
+    final JobsApi jobsApi = apiClient.getJobsApi();
+    final JobIdRequestBody body = new JobIdRequestBody().id(jobId);
+    final JobInfoLightRead jobInfo = AirbyteApiClient.retryWithJitter(
+        () -> jobsApi.getJobInfoLight(body),
+        "get job info");
+    return UUID.fromString(jobInfo.getJob().getConfigId());
+  }
+
+  /**
+   * Updates the Source from a sync job ID with the provided Configuration. Secrets and OAuth
+   * parameters will be masked when saving.
+   */
+  public void updateSource(final Long jobId, final Config config) {
+    final UUID connectionId = getConnectionIdFromJobId(jobId);
+
+    final ConnectionRead connection = AirbyteApiClient.retryWithJitter(
+        () -> apiClient.getConnectionApi().getConnection(new ConnectionIdRequestBody().connectionId(connectionId)),
+        "get connection");
+    final UUID sourceId = connection.getSourceId();
+
+    final SourceApi sourceApi = apiClient.getSourceApi();
+    final SourceRead source = AirbyteApiClient.retryWithJitter(
+        () -> sourceApi.getSource(new SourceIdRequestBody().sourceId(sourceId)),
+        "get source");
+
+    final SourceRead updatedSource = AirbyteApiClient.retryWithJitter(
+        () -> sourceApi
+            .updateSource(new SourceUpdate()
+                .sourceId(sourceId)
+                .name(source.getName())
+                .connectionConfiguration(Jsons.jsonNode(config.getAdditionalProperties()))),
+        "update source");
+
+    LOGGER.info("Persisted updated configuration for source {}. New config hash: {}.", sourceId,
+        Hashing.sha256().hashString(updatedSource.getConnectionConfiguration().asText(), StandardCharsets.UTF_8));
+
+  }
+
+  /**
+   * Updates the Destination from a sync job ID with the provided Configuration. Secrets and OAuth
+   * parameters will be masked when saving.
+   */
+  public void updateDestination(final Long jobId, final Config config) {
```

Comment:
Yeah, it's tricky because even though Sources/Destinations are equivalent in the DB, they're handled  as completely different in the java classes. Unfortunately there's tons of this kind of duplication in the java codebase for handling sources and destinations. All the methods for interacting with the classes are also type-specific, e.g. `.sourceId`, `.updateSource` etc.